File: bindings/typescript/packages/aster/src/trust/rcan.ts
/**
 * RCAN (Role-based Capability Access control Network) validation.
 *
 * Spec reference: Aster-trust-spec.md
 *
 * Evaluates caller roles against capability requirements (ROLE, ANY_OF, ALL_OF).
 * Also provides encode/decode for RCAN grant bytes (opaque until spec pins format).
 */

import type { CapabilityRequirement } from '../service.js';

/** Extract caller roles from attributes (comma-separated aster.role). */
export function extractCallerRoles(attributes: Record<string, string>): Set<string> {
  const roleStr = attributes['aster.role'] ?? '';
  if (!roleStr) return new Set();
  return new Set(roleStr.split(',').map(r => r.trim()).filter(Boolean));
}

/**
 * Evaluate a capability requirement against caller attributes.
 *
 * @returns true if the caller satisfies the requirement.
 */
export function evaluateCapability(
  requirement: CapabilityRequirement,
  callerAttributes: Record<string, string>,
): boolean {
  const roles = extractCallerRoles(callerAttributes);

  switch (requirement.kind) {
    case 'role':
      // Must have the single required role
      return requirement.roles.length > 0 && roles.has(requirement.roles[0]!);
    case 'any_of':
      // Must have at least one of the listed roles
      return requirement.roles.some(r => roles.has(r));
    case 'all_of':
      // Must have every listed role
      return requirement.roles.every(r => roles.has(r));
    default:
      return false;
  }
}

/**
 * Validate an RCAN grant (opaque bytes).
 * Currently a stub — non-empty grants accepted, empty bytes rejected.
 *
 * @returns [valid, reason]
 */
export function validateRcan(rcanBytes: Uint8Array): [boolean, string | undefined] {
  if (rcanBytes.length === 0) {
    return [false, 'empty RCAN grant'];
  }
  return [true, undefined];
}

/** Pass-through encoder for RCAN grant bytes. */
export function encodeRcan(data: Uint8Array): Uint8Array {
  return data;
}

/** Pass-through decoder for RCAN grant bytes. */
export function decodeRcan(data: Uint8Array): Uint8Array {
  return data;
}

File: bindings/python/aster/trust/rcan.py
"""
aster.trust.rcan -- rcan capability evaluation and grant serialization.

Provides:
- ``evaluate_capability``: runtime check of a caller's roles against a
  ``CapabilityRequirement`` (ROLE, ANY_OF, ALL_OF).
- Opaque encode/decode/validate helpers for rcan grant bytes (stub until
  upstream spec pins down the serialization format, §14.12).
"""

from __future__ import annotations

from aster.contract.identity import CapabilityKind, CapabilityRequirement


# ── Capability evaluation (Gate 3 -- method-level access control) ────────────


def _extract_caller_roles(caller_attributes: dict[str, str]) -> set[str]:
    """Extract the set of roles from caller admission attributes.

    The ``aster.role`` attribute may be:
    - A comma-separated string (e.g. ``"admin,editor"``).
    - Already a list (if the admission layer passed one through).

    Returns a set of stripped, non-empty role strings.
    """
    raw = caller_attributes.get("aster.role", "")
    if isinstance(raw, list):
        return {r.strip() for r in raw if r.strip()}
    if isinstance(raw, str):
        return {r.strip() for r in raw.split(",") if r.strip()}
    return set()


def evaluate_capability(
    requirement: CapabilityRequirement,
    caller_attributes: dict[str, str],
) -> bool:
    """Check whether *caller_attributes* satisfy *requirement*.

    Args:
        requirement: The ``CapabilityRequirement`` from the service or method
            contract (``ROLE``, ``ANY_OF``, or ``ALL_OF``).
        caller_attributes: The admission attributes dict carried on
            ``CallContext.attributes``.  The key ``aster.role`` holds the
            caller's role(s).

    Returns:
        ``True`` if the caller satisfies the requirement, ``False`` otherwise.
    """
    caller_roles = _extract_caller_roles(caller_attributes)

    if requirement.kind == CapabilityKind.ROLE:
        # ROLE: caller must have the single required role.
        if not requirement.roles:
            return True  # vacuously satisfied -- no role demanded
        return requirement.roles[0] in caller_roles

    if requirement.kind == CapabilityKind.ANY_OF:
        # ANY_OF: caller must have at least one of the listed roles.
        if not requirement.roles:
            return True
        return bool(caller_roles & set(requirement.roles))

    if requirement.kind == CapabilityKind.ALL_OF:
        # ALL_OF: caller must have every listed role.
        if not requirement.roles:
            return True
        return set(requirement.roles) <= caller_roles

    # Unknown kind -- fail closed.
    return False


# ── Opaque grant helpers (stub) ─────────────────────────────────────────────


def validate_rcan(rcan_bytes: bytes) -> tuple[bool, str | None]:
    """Validate an rcan grant.

    Phase 12: all non-empty grants are accepted (opaque).  Empty bytes -> invalid.
    """
    if not rcan_bytes:
        return False, "rcan grant must not be empty"
    # TODO: implement full rcan validation once upstream format is specified (§14.12)
    return True, None
